src/environmental/water_quality/surface_water_quality.py

```diff
@@ -0,0 +1,304 @@
+    # Merge the fetched water quality data with site geometries to add geometric information
+    sites_geometry = surface_water_sites[["site_id", "geometry"]]
+    water_quality_with_geom = water_quality_data.merge(sites_geometry, how="left", on="site_id", validate="many_to_one")
+    # Convert the merged result into a GeoDataFrame and reset the index
+    water_quality_with_geom = gpd.GeoDataFrame(water_quality_with_geom).reset_index(drop=True)
+    return water_quality_with_geom
+
+
+def get_surface_water_quality_from_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Retrieve surface water quality data from the database for the specified catchment area.
+
+    Parameters
+    ----------
+    engine : Engine
+        The engine used to connect to the database.
+    catchment_area : gpd.GeoDataFrame
+        A GeoDataFrame representing the catchment area.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        A GeoDataFrame containing the retrieved surface water quality data for the requested catchment area.
+    """
+    # Extract the geometry of the catchment area and its corresponding CRS
+    catchment_polygon = catchment_area["geometry"][0]
+    catchment_crs = catchment_area.crs.to_epsg()
+    # Query to retrieve surface water quality data that intersects with the catchment polygon
+    command_text = """
+        SELECT *
+        FROM surface_water_quality AS swq
+        WHERE ST_Intersects(swq.geometry, ST_GeomFromText(:catchment_polygon, :catchment_crs));
+        """
+    query = text(command_text).bindparams(
+        catchment_polygon=str(catchment_polygon),
+        catchment_crs=str(catchment_crs)
+    )
+    # Execute the query and create a GeoDataFrame from the result
+    swq_data = gpd.GeoDataFrame.from_postgis(query, engine, geom_col="geometry")
+    return swq_data
+
+
+def get_surface_water_quality_not_in_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Retrieve surface water quality data from ECAN for the specified catchment area that is not already present
+    in the existing database.
+
+    Parameters
+    ----------
+    engine : Engine
+        The engine used to connect to the database.
+    catchment_area : gpd.GeoDataFrame
+        A GeoDataFrame representing the catchment area.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        A GeoDataFrame containing surface water quality data from ECAN for the requested catchment area that is
+        not already present in the existing database.
+    """  # noqa: D400
+    # Retrieve existing surface water quality data from the database for the requested catchment area
+    water_quality_exist = get_surface_water_quality_from_db(engine, catchment_area)
+    # Fetch surface water quality data from ECAN for the requested catchment area
+    water_quality_new = get_surface_water_quality_with_geom(engine, catchment_area)
```

Comment:
This is feasible, but it would stop fetching the latest data for areas already covered. That's why I mentioned the current design principle in the notes section—it's open for discussion. I believe we should always visualize the latest data (even though it might not be updated often). What are your thoughts? Or if you have another approach, let me know.

Reply:
Let's discuss in person, I think I agree with you but I want to understand the implications more in depth. Implications like the number of requests we're sending etc.